all/HOPE-3--Dependency-Injection.md

```diff
@@ -78,45 +78,52 @@ It should also be possible to specify different dependency providers for differe
 
 ```python
     # Multiple interfaces using one provider
-    @hug.provides("mysql", interfaces=['http', 'websocket'])
+    @hug.provide("mysql", interfaces=['http', 'websocket'])
     def mysql_web():
         ...
 
 
     # For a provider for a single interface, this shorthand can be used
-    @hug.cli.provides("mysql")
+    @hug.cli.provide("mysql")
     def mysql_cli():
         ...
 
 
     # The following would also be a valid way to apply to multiple interfaces
-    @hug.http.provides("mysql")
-    @hug.cli.provides("mysql")
+    @hug.http.provide("mysql")
+    @hug.cli.provide("mysql")
+    def mysql():
+        ...
+
+    # Or the following would also be a valid way to apply to multiple interfaces
+    @hug.http.provide
+    @hug.cli.provide
     def mysql():
         ...
 ```
 
 If you only want to enable provider usage locally, and don't want to allow it to be overridden, you can explicitly set it to stay unnamed:
 
 ```python
-    @hug.provides(name=False)
+    @hug.provide(name=False)
     def mysql_connection(host, port, ...):
         ...
 ```
 
 ### Using a dependency
 
-Dependencies will be pulled into individual endpoints via a `requires(` function, that will be called within function/method defaults:
+Dependencies will be pulled into individual endpoints via a `require` decorator, and provided to the calling function
+via an eponymous argument
 
 ```python
-from hug import requires
 
-@hug.http()
-def hello_world(mysql=requires('mysql')):
+@hug.http
+@hug.require.mysql
+def hello_world(mysql):
```

Comment:
Here are the motivating thoughts I had in refactoring this:
- get rid of string value connascence
- also, in contrast to string connascence, we could get IDE tab over completion (implementation and IDE dependent)
- would have functional closure to accommodate monkey patching on dynamic changes. The string default form evaluates on import time only
- get rid of a "fake arg". It's a little odd to use a signature to define an arg which is only ever intended to take a default value. Overriding the default would be a violation of our intended contract, but this is at odds with how Python generally models data.
- pytest has high traction and a pretty good "magic" API. I think we can expect users to understand magic that would be similar or the same to pytest. This is close to the pytest fixture magic, although using the decorator makes this pattern more explicit than the pytest fixture pattern

Happy to discuss further!

Reply:
I think points 3 and 4 (1 indexed) counteract each-other a bit as providing a parameter that defaults to a dependency is an acceptable behavior for test cases. A test case could either provide the arguments that the dependency requires and leave parameter alone or it could ignore the dependency injection arguments and directly override the parameter with the data the dependency would provide. (This does make me think of a separate point for this HOPE, there should be a warning if a dependency parameter is provided as well as the parameter to which the dependency return is assigned)

For point 1 I can see the value in reducing the amount of "strings as code",  but one way or the other the named `identifier` has to be referenced indirectly through the API. To pull in point 2 I think we could potentially have `requires` be a class such that you could call:
```python
@hug.http()
def hello_world(mysql=requires.mysql()):
```
I actually like this form a bit better than the initial suggestion at line 31 in that it is more clear that `*args, **kwargs` goes directly to the dependency function.

Per point 5 I know pytest uses the magic parameter name, but I've seen it cause confusion for multiple new coders, and personally I think testing code has different behavior standards as by its nature it is performing meta logic. That being said HUG itself has liberally used this format in the past so I don't consider it as unacceptable when designing new HUG features. I think it is reasonable behavior, but that's not the same saying it is the cleanest form for designing new functionality as in this HOPE.

I'm quite open to changing my thoughts here, just want to make sure I'm explaining myself clearly. I definitely value having multiple viewpoints on the HOPE!

Reply:
I dunno, it's apples to oranges on the magic end. I don't think "magic" is necessarily a bad thing, but it's probably better to leverage magic via a decorator when it's needed. It's explicit and well understood that a decorator is allowed to change callable flow. Also it's the only kind of magic we see provided in the standard library. Thinking of `functools.total_ordering` and `contextlib.context_manager`. Both very "magic" but in a fairly explicit and understandable way.

Thinking the default arg approach
- Raises questions about control flow (why a default arg? and who is calling hello_world? and where is the contract of hello_world specified fully? )
- Raises questions about what requires.mysql is doing
- Is harder to remember as a pattern

Thinking the decorator approach
- Raises a question about the `hug.provide` decorator and how it works

Not totally understanding your point about points 3 and 4 counteracting each other. Point 3 is the import time issue, and I'm not sure how you'd get around it. It's going to evaluate as an expression at import time, and the only way to get around it is to return a callable as a result of `requires.mysql()`, e.g. `requires.mysql()()` would give the value. But that would be fairly confusing to end users IMO.

I think both approaches raise questions for beginners. Most interested in the approach that helps people pose a question to the right place (in code, ideally), and is easiest to remember as a pattern once used.

Also - what unit test framework is using default arg setting as an injection mechanism? I'd like to take a look.

Reply:
As I've thought this through I agree the level of magic is fairly similar in either pattern. I agree that using a decorator to change the call-time behavior of the function is the more recognizable pattern. The decorator way the HUG code doesn't need to reach inside the function using require to associate the parameter with the dependency injection.

I'm curious to get @timothycrosley's thoughts, but my understanding of the `def function(db = requires('mysql'))` is that a callable was stored in the default parameter for `db` and that HUG would look for the callable in the function's parameter defaults when handing off execution to the endpoint. Anyhow I agree that it is not an immediately recognizable behavior pattern.

I'm not aware of any unit test framework using default arguments to parameters as an injection mechanism. I was talking about pytest same as your first comment's 5th bullet point. The part I particularly found confusing in pytest is that it had to introspect the functions you wrote to determine the name of parameters and match fixtures to those names. However when there's a decorator directly attached to the function (and especially if the decorator name matches the parameter name) it reduces the scope of investigation to track down the source of data to the parameter.

tl;dr: I agree with moving to a decorator for `require`